baseband/guppi/base.py

```diff
@@ -0,0 +1,338 @@
+        """Get frame by writing the header to disk and mapping its payload.
+
+        The header is written to disk immediately, but the payload is mapped,
+        so that it can be filled in pieces, by setting slices of the frame.
+
+        Parameters
+        ----------
+        header : `~baseband.guppi.GUPPIHeader`
+            Written to disk immediately.  Can instead give keyword arguments to
+            construct a header.
+        **kwargs
+            If ``header`` is not given, these are used to initialize one.
+
+        Returns
+        -------
+        frame: `~baseband.guppi.GUPPIFrame`
+            By assigning slices to data, the payload can be encoded piecewise.
+        """
+        if header is None:
+            header = GUPPIHeader.fromvalues(**kwargs)
+        header.tofile(self.fh_raw)
+        payload = GUPPIPayload.fromfile(self.fh_raw, memmap=True,
+                                        header=header)
+        return GUPPIFrame(header, payload)
+
+
+class GUPPIStreamBase(VLBIStreamBase):
+    """Base for GUPPI streams."""
+
+    _sample_shape_maker = GUPPIPayload._sample_shape_maker
+
+    def __init__(self, fh_raw, header0, squeeze=True, subset=()):
+
+        # GUPPI sample positions are determined by "packets", the bytesize of
+        # which are given in the header.
+        self._packets_per_frame = (
+            (header0.payload_nbytes -
+             header0.overlap * header0.bits_per_complete_sample // 8) //
+            header0['PKTSIZE'])
+
+        super(GUPPIStreamBase, self).__init__(
+            fh_raw=fh_raw, header0=header0, sample_rate=header0.sample_rate,
+            samples_per_frame=header0.samples_per_frame,
+            unsliced_shape=header0.sample_shape, bps=header0.bps,
+            complex_data=header0.complex_data, squeeze=squeeze, subset=subset,
+            fill_value=0.)
+
+
+class GUPPIStreamReader(GUPPIStreamBase, VLBIStreamReaderBase):
+    """DADA format reader.
+
+    Allows access to DADA files as a continuous series of samples.
+
+    Parameters
+    ----------
+    fh_raw : filehandle
+        Filehandle of the raw DADA stream.
+    squeeze : bool, optional
+        If `True` (default), remove any dimensions of length unity from
+        decoded data.
+    subset : indexing object or tuple of objects, optional
+        Specific components of the complete sample to decode (after possibly
+        squeezing).  If a single indexing object is passed, it selects
+        polarizations.  With a tuple, the first selects polarizations and the
+        second selects channels.  If the tuple is empty (default), all
+        components are read.
+    """
+    def __init__(self, fh_raw, squeeze=True, subset=()):
+        fh_raw = GUPPIFileReader(fh_raw)
+        header0 = GUPPIHeader.fromfile(fh_raw)
+        super(GUPPIStreamReader, self).__init__(fh_raw, header0,
+                                                squeeze=squeeze, subset=subset)
```

Comment:
My own feeling is that almost everything would work automatically if we were to reset `self.samples_per_frame` here accounting for the overlap.